### ISSUE TITLE
negative retval not working in filter

### ISSUE THREAD

**Issue by brendangregg** [iovisor/bpftrace#668]
This should only count the retval if it is > 0:

```
# bpftrace -e 'kr:sock_sendmsg,kr:sock_recvmsg /retval > 0/ { @[retval]++ }'
[...]
@[120]: 44
@[40]: 51
@[216]: 57
@[112]: 82
@[32]: 201
@[4294967285]: 4386
```

4294967285 is -11 (EAGAIN). But we shouldn't be counting it, as it should have been filtered via the /retval > 0/ test.

**Comment by mmarchini** [comment]
Might be related to #585

**Comment by fbs** [comment]
I think this is because `sock_sendmsg` returns an int32.  An 32bit value gets automatically *zero* extended to 64 bits by the CPU. So the register will hold `0x00..FF..`  instead of `0xFF...` for -1.

As a test this seems to fix it for the `binops`:

```diff
diff --git a/src/ast/codegen_llvm.cpp b/src/ast/codegen_llvm.cpp
index 04f5ac7..ed2bfd9 100644
--- a/src/ast/codegen_llvm.cpp
+++ b/src/ast/codegen_llvm.cpp
@@ -157,9 +157,10 @@ void CodegenLLVM::visit(Builtin &builtin)
       offset = arch::arg_offset(arg_num);
     }

-    AllocaInst *dst = b_.CreateAllocaBPF(builtin.type, builtin.ident);
+    AllocaInst *dst = b_.CreateAllocaBPF(b_.getInt32Ty(), builtin.ident);
     Value *src = b_.CreateGEP(ctx_, b_.getInt64(offset * sizeof(uintptr_t)));
-    b_.CreateProbeRead(dst, builtin.type.size, src);
+    b_.CreateProbeRead(dst, 4, src);
     expr_ = b_.CreateLoad(dst);
     b_.CreateLifetimeEnd(dst);
   }
@@ -843,8 +844,8 @@ void CodegenLLVM::visit(Binop &binop)
     rhs = expr_;

     // promote int to 64-bit
-    lhs = b_.CreateIntCast(lhs, b_.getInt64Ty(), false);
-    rhs = b_.CreateIntCast(rhs, b_.getInt64Ty(), false);
+    lhs = b_.CreateIntCast(lhs, b_.getInt64Ty(), true);
+    rhs = b_.CreateIntCast(rhs, b_.getInt64Ty(), true);

     switch (binop.op) {
       case bpftrace::Parser::token::EQ:    expr_ = b_.CreateICmpEQ (lhs, rhs); break;
```
(do a 32 bits read and then sign extend it to a 64bit in the binop)

 (tested with `sudo ./src/bpftrace -e 'kr:sock_sendmsg,kr:sock_recvmsg /retval > 0/ { @x[retval]++ } kr:sock_sendmsg,kr:sock_recvmsg /retval < 0/ {  @t[retval]++ }'`, map t has `4294967285` as expected, map x values stay under 2k)

But this will break a lot of other things. 

I looked for a way to be able to determine the return type but it doesn't seem like the kernel has that (not for kprobes at least)

Introducing `retval32` could be an option, but feels ugly

**Comment by mmarchini** [comment]
bcc seems to treat `retval` as a 32-bit signed int. The kernel stores it as an `unsigned long` though (https://github.com/torvalds/linux/blob/master/arch/x86/include/asm/ptrace.h#L72). We could follow bcc lead, but I think the best solution might be to allow users to cast to integer types and have `retval` as unsigned int by default.

**Comment by brendangregg** [comment]
I don't mind `retval` always being u64 and needing to cast it as int32 when necessary. Just can't cast ints at the moment. :)

**Comment by danobi** [comment]
I think @fbs  fixed this in #922 . Please reopen if I'm mistaken
